File: app/src/main/java/org/wikipedia/util/ImageUrlUtil.java
package org.wikipedia.util;

import android.net.Uri;
import android.support.annotation.NonNull;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

public final class ImageUrlUtil {
    private static Pattern WIDTH_IN_IMAGE_URL_REGEX = Pattern.compile("/(\\d+)px-");

    @NonNull
    public static Uri getUrlForSize(@NonNull Uri uri, int size) {
        return Uri.parse(getUrlForSize(uri.toString(), size));
    }

    @NonNull
    public static String getUrlForSize(@NonNull String original, int size) {
        Matcher matcher = WIDTH_IN_IMAGE_URL_REGEX.matcher(original);
        if (matcher.find() && Integer.parseInt(matcher.group(1)) > size) {
            return original.replace(matcher.group(1), Integer.toString(size));
        }
        return original;
    }

    private ImageUrlUtil() {
    }
}

File: app/src/main/java/org/wikipedia/util/ImageUrlUtil.kt
package org.wikipedia.util

object ImageUrlUtil {

    private val WIDTH_IN_IMAGE_URL_REGEX = """/(page\d+-)?(\d+)(px-)""".toPattern()

    fun getUrlForPreferredSize(original: String, size: Int): String {
        val matcher = WIDTH_IN_IMAGE_URL_REGEX.matcher(original)
        return if (matcher.find() && matcher.group(2)!!.toInt() != size) {
            matcher.replaceFirst("/${matcher.group(1).orEmpty()}$size$3")
        } else {
            original
        }
    }

    fun insertLangIntoThumbUrl(original: String, lang: String): String {
        val matcher = WIDTH_IN_IMAGE_URL_REGEX.matcher(original)
        return if (matcher.find() && matcher.group(2)!!.toInt() > 0) {
            matcher.replaceFirst("/${matcher.group(1).orEmpty()}lang$lang-${matcher.group(2).orEmpty()}$3")
        } else {
            original
        }
    }

    fun isGif(url: String?): Boolean {
        return url?.lowercase()?.endsWith(".gif") == true
    }
}
